MobilityData/mobility-feed-api
# feat: tie API usage to user account

**Summary:**

This PR logs the user ID to cloud logging. The ID doesn't contain any sensitive data. In addition, the user's information is added to the context of each request.

**Expected behavior:** 
The API produces logs structured with relevant information on each call, example:
```
{
  "insertId": "663be1bb000ea4abbd1fffe3",
  "jsonPayload": {
    "traceSampled": false,
    "user_id": "HERE the ID",
    "httpRequest": {
      "latency": 0.16521739959716797,
      "status": 200,
      "userAgent": "FULL USER AGENT",
      "remoteIp": "HERE IP",
      "requestMethod": "GET",
      "responseSize": 1997,
      "requestUrl": "https://api-dev.mobilitydatabase.org/v1/gtfs_feeds/mdb-100/datasets",
      "protocol": "HTTP/1.1"
    },
    "trace": "HERE TRACE INFORMATION",
    "textPayload": "API Access Log",
    "spanId": "HERE SPAN ID",
    "jsonPayload": null
  },
..._other GCP standard logs fields_...
}
```

Explain and/or show screenshots for how you expect the pull request to work in your testing (in case other devices exhibit different behavior).

**Testing tips:**

Local testing:
- Checkout branch
- Update Python dependencies: `pip3 install -r api/requirements.txt` and 
- Reset your local db: `./scripts/docker-localdb-rebuild-data.sh --populate-db`
- Start API: `./scripts/api-start.sh
- Execute any API REST call, example:
```
curl --location 'http://localhost:8080/v1/feeds' \
--header 'Accept: application/json'
```
- Expect logs in the console formatted like:
```
{"user_id": null, "httpRequest": {"request method": "GET", "requestUrl": "http://localhost:8080/v1/feeds", "status": 200, "responseSize": 967618, "userAgent": "PostmanRuntime/7.37.0", "remoteIp": "127.0.0.1", "serverIp": "127.0.0.1", "latency": 0.4000420570373535, "protocol": "HTTP/1.1"}, "trace": "", "spanId": null, "traceSampled": false, "textPayload": "API Access Log", "jsonPayload": null}
```
QA testing
- Deploy branch with GitHub action `Deploy Feeds API - QA`
- Login to QA UI environment from https://qa.mobilitydatabase.org/sign-in
- Generate an access token
-  Execute a REST call, example:
```
curl --location 'https://api-qa.mobilitydatabase.org' \
--header 'Accept: application/json' \
--header 'Authorization: Bearer <YOUR TOKEN HERE>'
```
- Explore Google Cloud logging and expect a log message formatted like:
```
{
  "insertId": "663cefe3000690ef069aa184",
  "jsonPayload": {
    "jsonPayload": null,
    "traceSampled": false,
    "textPayload": "API Access Log",
    "user_id": "securetoken.google.com/mobility-feeds-qa:HERE_USER_ID",
    "spanId": "7712122025680621213",
    "httpRequest": {
      "status": 200,
      "userAgent": "python-requests/2.31.0",
      "requestMethod": "GET",
      "requestUrl": "https://api-qa.mobilitydatabase.org/v1/gtfs_feeds",
      "serverIp": "34.111.32.89",
      "latency": 0.009491205215454102,
      "responseSize": 358,
      "protocol": "HTTP/1.1",
      "remoteIp": "13.88.99.132"
    },
    "trace": "projects/mobility-feeds-qa/traces/58b64443d53068a65f69bd7086b18e19"
  },
  "resource": {
    "type": "cloud_run_revision",
    "labels": {
...labels
    }
  },
  "timestamp": "2024-05-09T15:46:43.430319Z",
  "labels": {
    "instanceId": "00ff"
  },
  "logName": "projects/mobility-feeds-qa/logs/run.googleapis.com%2Fstderr",
  "receiveTimestamp": "2024-05-09T15:46:43.653381420Z"
}
```
Please make sure these boxes are checked before submitting your pull request - thanks!
